bam_addrprg.c

```diff
@@ -0,0 +1,471 @@
+    }
+    return tmp;
+}
+
+// These are to be replaced by samtools header parser
+// Extracts the first @RG line from a string.
+static char* get_rg_line(const char* text, size_t* last)
+{
+    const char* rg = text;
+    if (rg[0] != '@' || rg[1] != 'R' || rg[2] != 'G' ) {
+        if ((rg = (const char*)strstr(text,"\n@RG")) == NULL) {
+            return NULL;
+        }
+        rg++;//skip initial \n
+    }
+    // duplicate the line for return
+    char* line;
+    char* end = strchr(rg, '\n');
+    if (end) {
+        line = strndup(rg,(end-rg));
+        *last = end - rg;
+    } else {
+        line = strdup(rg);
+        *last = strlen(rg);
+    }
+    return line;
+}
+
+// Given a @RG line return the id
+static char* get_rg_id(const char* input)
+{
+    assert(input!=NULL);
+    char* line = strdup(input);
+    char *next = line;
+    char* token = strsep(&next, "\t");
+    token = strsep(&next,"\t"); // skip first token it should always be "@RG"
+    while (next != NULL) {
+        char* key = strsep(&token,":");
+        if (!strcmp(key,"ID")) {
+            char* retval = strdup(token);
+            free(line);
+            return retval;
+        }
+        token = strsep(&next,"\t");
+    }
+    free(line);
+    return NULL;
+}
+
+// Confirms the existance of an RG line with a given ID in a bam header
+static bool confirm_rg( const bam_hdr_t *hdr, const char* rgid )
+{
+    assert( hdr != NULL && rgid != NULL );
+
+    char *ptr, *start;
+    bool found = false;
+    start = ptr = strndup(hdr->text, hdr->l_text);
+    while (ptr != NULL && *ptr != '\0' && found == false ) {
+        size_t end = 0;
+        char* line = get_rg_line(ptr, &end);
+        if (line == NULL) break; // No more @RG
+        char* id;
+        if (((id = get_rg_id(line)) != NULL) && !strcmp(id, rgid)) {
+            found = true;
+        }
+        free(id);
+        free(line);
+        ptr += end;
+    }
+    free(start);
+    return found;
+}
+
+static char* get_first_rgid( const bam_hdr_t *hdr )
+{
+    assert( hdr != NULL );
+    char *ptr, *start;
+    char* found = NULL;
+    start = ptr = strndup(hdr->text, hdr->l_text);
+    while (ptr != NULL && *ptr != '\0' && found == NULL ) {
+        size_t end = 0;
+        char* line = get_rg_line(ptr, &end);
+        if ( line ) {
+            found = get_rg_id(line);
+        } else break;
+        free(line);
+        ptr += end;
+    }
+    free(start);
+    return found;
+}
+
+static void usage(FILE *fp)
+{
+    fprintf(fp,
+            "Usage: samtools addreplacerg [options] [-r <@RG line> | -R <existing id>] <input.bam> <output.bam>\n"
+            "\n"
+            "Options:\n"
+            "  -r STRING @RG line text\n"
+            "  -R STRING ID of @RG line in existing header to use\n"
+            "  -m MODE   Set the mode of operation from one of overwrite_all, orphan_only [overwrite_all]\n"
+            "  -l INT    Set compression level, from 0 (uncompressed) to 9 (best)\n");
```

Comment:
Now that we have the `-O` and `--output-fmt-options`, we shouldn't need `-l level` options any more.